### DISCUSSION TITLE
How do I NOT escape special characters from the {{clipbboard}}?

### DISCUSSION THREAD

**Discussion by wealthychef1** [Taitava/obsidian-shellcommands#224] [Q&A] [answered]
When I have a path name on the clipboard, all the slashes and spaces get escaped by a backslash like this:  `\/Users\/rcook\/Library\/Mobile\ Documents\/iCloud\~com\~omnigroup\~OmniGraffle\/Documents\/2021\-2022\ Timelines\.graffle`

I am trying to form a URL from this (maybe there is already a plugin for file URLs?) but I can't because of the slashes here, makes it way harder.  I just want to put angle brackets around the pathname and it would work with spaces.  
So how do I NOT escape special chars?  Thanks!

**Comment by wealthychef1** [comment]
I have learned that I can achieve what I want with option-click-drag of a file to a document.  However, I have hit this problem before in other contexts, so I'd still like to know how to escape the escapes!   Thanks

**Comment by Taitava** [comment, marked as answer]
Hi! If you really want to use unescaped values, add an exclamation mark `!` to the variable: `{{!clipboard}}`. In case you want to read more about escaping, there's a small documentation page for it: https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values

May I ask about what is the precise shell command where the escaping causes problems? I'm interested to learn use-cases that are new to me. Usually I'd say that **unescaped values can be dangerous to use**, but I don't know the situation here. The backslashes should be removed by the shell, at least if you don't wrap the `{{clipboard}}` variable in quotes (when your shell is Bash/Dash/Zsh. For PowerShell, variables can be wrapped in quotes, too). But maybe I could see the situation better if I'd see the whole command. 🙂 

There's still one problem left: Currently the _Shell commands_ plugin does not offer a way to do [URL encoding](https://www.w3schools.com/tags/ref_urlencode.asp), so some special characters (such as spaces) in the clipboard content may break your URL. [I'm trying to come up with some kind of url encoding plan here](https://github.com/Taitava/obsidian-shellcommands/discussions/95#discussioncomment-2125652). Maybe you can use some shell command to do URL encoding.

**Reply by wealthychef1** [reply]
>  [I'm trying to come up with some kind of url encoding plan here](https://github.com/Taitava/obsidian-shellcommands/discussions/95#discussioncomment-2125652). Maybe you can use some shell command to do URL encoding.

I was just going to go with `[LINK LABEL](<file:///path/to/file/including spaces/ and ~wierd stuff>)`. The <> syntax protects all the junk in there.

**Reply by Taitava** [reply]
> The <> syntax protects all the junk in there.

Sorry, I did not understand this part 🙂. How does it protect something?

**Reply by wealthychef1** [reply]
In Obsidian, you can put <> around your URL and it does not need % character escape sequences.

**Reply by Taitava** [reply]
That's nice to know! 😎

**Comment by wealthychef1** [comment]
Thanks for the trick!  I will certainly use ! for my little hacks going forward.  And the documentation link is very helpful. 

As for a use case, when I use the command `echo -n {{clipboard}}` it results in something like `\/Users\/rcook\/Library\/Mobile\ Documents\/iCloud\~com\~omnigroup\~OmniGraffle\/Documents\/2021\-2022\ Timelines\.graffle`.  I don't need that aggressive level of escapes.  What's wrong with forward slashes and tildes and spaces?  Maybe in the world of markdown those are sketchy chars?  Anyhow, for my use, I want no escapes.  It just adds confusing to have another hand in there modifying things.

**Reply by Taitava** [reply]
> Thanks for the trick! I will certainly use ! for my little hacks going forward.

You are welcome! However, I still recommend you to try to get your commands working with escaping.

> What's wrong with forward slashes and tildes and spaces?

- Forward slashes: I'm not sure if there's anything bad with them, maybe not. The escaping system is based on the idea that it's better to escape all special characters (so any character that is not a number, a letter nor an underscore `_`), because I don't know all the possible special meanings each character may have in different situations in shells. Maybe forwardslashes never have any special meaning? They are escaped just in case.
- A tilde is a special character which - when appearing alone - means the user's home directory. E.g. `echo ~` outputs `/home/jarkko` for me, while `echo \~` outputs `~`. The escaping system does not check if `~` appears alone or not, it escapes it always, just like all other special characters.
- Unescaped spaces are often interpreted as argument separators, but not when using `echo`, which reads the whole rest of the line after any options (so, after `-n` in this case). If you'd use for example `mkdir {{!clipboard}}` and your clipboard happened to contain a text that contains a space, e.g. `my folder`, the command would become `mkdir my folder`, which would create **two** folders instead of one, `my` and `folder`. However, if you use the recommended escaped form `mkdir {{clipboard}}`, then the command would become `mkdir my\ folder`, which creates a single folder: `my folder`. Note that the shell removes the backslash `\` when it interprets the space character.

For your situation, there's probably some unnecassary escaping, but if you can tweak your shell command to work well with the escaped variable values, then the backslahes should disappear from the output. I still recommend you to try to use the escaping, as there might become a time when your variable values contain other special characters, such as `>` which can be dangerous if it's unescaped. For example, if your clipboard contains the text `Note.md > Something` and you use a shell command `echo -n {{!clipboard}}`, it becomes `echo -n Note.md > Something`. The greater than character `>` will cause a text `Note.md` to be written to a file named `Something`. However, an escaped `echo -n {{clipboard}}` would become `echo -n Note\.md\ \>\ Something`, in which case `\>` is escaped, meaning that it will be used as a literal `>` part of the text that should be outputted.

> Maybe in the world of markdown those are sketchy chars?

It's not related to markdown. The escape characters are meant for the shell that interprets your shell commands, and the purpose is to tell the shell that these special characters should be interpreted as **literal** characters, removing any possible special meanings the characters could otherwise have in the shell. If the output of the shell command contains the backslash `\` escape characters, it's not intended. The shell should remove the backslashes when it interprets the shell command, but there are some exceptions when they are not removed, e.g. when text is wrapped in quotes (`'` or `"`), and - it seems - when using the `echo` command **in your shell**. In some shells, `echo` removes backslashes, at least in Bash. **What is your shell?** From the fact that your shell command contains backslashes `\` as escape characters, I know that your shell is either Bash, Dash or Zsh. Which one is it? I can then learn to understand better the differences of these shells, as I'm currently only familiar with Bash from those three.

> Anyhow, for my use, I want no escapes. It just adds confusing to have another hand in there modifying things.

Yea, it's confusing, as the reasons for the escaping can be unknown at the beginning, and because there are situations where the escape characters do not disappear. It just takes some time to learn how to cope with it.

## Something to try

You could try these shell commands with escaping:
- `echo -n -e {{clipboard}}`
- `printf {{clipboard}}`

**Reply by wealthychef1** [reply]
>Forward slashes: I'm not sure if there's anything bad with them, maybe not. The escaping system is based on the idea that it's better to escape all special characters (so any character that is ... they are escaped just in case...   I still recommend you to try to get your commands working with escaping.

I think you are assuming that cleaning up all your backslashes is trivial compared to removing spaces, but actually it just moves the problem and creates a new special character to worry about, which is actually harder to remove because backslash is itself a special shell character.  For most cases, I understand shell escaping and quoting just fine.  As someone who has used bash and unix for years, having a program try to escape its output like this is not helpful.  You have to assume you know my use cases, and you can't possibly know.  Just trust the shell.  It's there for a reason.  


>it's confusing, as the reasons for the escaping can be unknown at the beginning, 

I still don't know why you are doing this honestly.  Most people who use the shell are able to deal with quoting. If I need backslashes, I'll add them myself.  :-) Your idea of special characters seems arbitrary and unusual.  '/' is not a special character in any context I know of.  But if it is, then the USER should decide to escape it, not some algorithm which cannot understand the context the output will be used in.  Generic is better!  

>echo -n -e {{clipboard}}

My man page for echo does not have the -e option.  Sounds like a variant.  Your quoting method adds extra assumptions and steps to my workflow.  I have the ! trick and that's what I'll use personally.  I suggest you make the ! behavior the default ("don't change my text please") and offer "automatic quoting" for anyone that wants your version.  :-). 

>The escape characters are meant for the shell that interprets your shell commands, and the purpose is to tell the shell that these special characters should be interpreted as literal characters

DO NOT WANT!  Leave my output alone.  :-). I like the shell and how it interprets things.  Sometimes those interpolations are very desirable.  The tilde is a great example of such a case where I don't want to escape that handy construct!  
Of course there are times where I want to, and I will do so.  But it won't be by leaving the ! off, it will be by using the shell properly...

**Reply by Taitava** [reply]
> I think you are assuming that cleaning up all your backslashes is trivial compared to removing spaces,

Yes, I do assume that cleaning up the backslashes is trivial, as I assume your shell does it for you. But as it seems that your shell does not clean the backslashes, I'd like to know more about this situation: what is your shell and your operating system? Maybe it could help me understand why the backslashes are left uncleaned. 

> but actually it just moves the problem and creates a new special character to worry about, which is actually harder to remove because backslash is itself a special shell character.

Sorry, I did not understand this. Can you try to explain it again?

> For most cases, I understand shell escaping and quoting just fine. As someone who has used bash and unix for years, having a program try to escape its output like this is not helpful.

Ok, good to know that you are well experienced with shell escaping 👍. Then you are well prepared to use the unescaped `!` way. I'd still like us to continue inspecting why the escaping does not work in your use case. You don't have to use the escaping if you don't want to, I'd just like to find a solution to make it work, just to learn myself and to find out if there's anything for me to improve in the plugin.

> You have to assume you know my use cases, and you can't possibly know.

Yes, the escaping that the plugin uses, does assume that the variable values are used in a certain way, i.e. not wrapped in any quotes (except in PowerShell, but that's not the case here). Apparently, there's more assumptions, that I'm yet unaware of, because you are having this problem even when you don't use any quotes. I'd like to know what it is that causes the backslashes not to disappear.

I don't have extensive experience on shell escaping myself. I wanted to come up with an escaping system that is safe to use even when the variable values contain characters the user who wrote the shell command did not expect. It's really valuable to get feedback about the escaping 🙂.

- I'd still like to know what shell you are using. Maybe I could then better understand why the backslashes were not removed by the shell.
- Can you still try the `printf {{clipboard}}` command? You don't have to take it into use, but I'd like to know if it produces output that has backslashes removed or not.

I want to learn different use cases people have with this plugin, and so I'll definitely want to hear more about how you are using it.

> I suggest you make the ! behavior the default ("don't change my text please") and offer "automatic quoting" for anyone that wants your version. :-).

Changing the default behaviour is quite unlikely to happen. I chose the escaping to be the default behaviour because in my opinion, variable values should be safe to use. In my opinion, it's more preferable that variable values in some cases leave visible backslashes, if it can avoid the more drastic problems that unescaped special characters, such as `"`, `;` and `>` may cause: unexpected "jail break" from a string, unexpected end of command (and accidentally beginning a new command), and unexpected file writing. If in the future it would become apparent that the escaping would cause more problems than what it solves, then I might consider changing the default.

I have thought about the "don't change my text please" aspect myself, too, and if I would have found another way to insert variable values safely into the shell commands, I would definitely have used it. I don't like to alter users' values. I just haven't found a better way to ensure safety. (For example, I'd love it if there would be a way to use parameterised shell commands, like there is a way to use parameterised queries in SQL in order to avoid [SQL injection](https://www.w3schools.com/sql/sql_injection.asp), which is technically very similar to this escaping problem here, although in the case of Obsidian, there is no way for unknown users being able to execute untrusted commands, like there is in web.) If you have ideas on how to pass variable values to shell commands safely without using escaping, please let me know, I'm willing to learn! 🙂 

> Most people who use the shell are able to deal with quoting. If I need backslashes, I'll add them myself. :-)

If a user wants to select some text (e.g. a paragraph) in a note and then use it as a parameter to a shell command via the [{{selection}} variable](https://publish.obsidian.md/shellcommands/Variables/%7B%7Bselection%7D%7D), I would not expect the user to first go add some escaping characters manually in front of every character that may break their shell command. I don't think it would make sense to do it in a note file, which is originally not thought to be used as part of some shell command.

> Your idea of special characters seems arbitrary and unusual. '/' is not a special character in any context I know of. But if it is, then the USER should decide to escape it

Yes, `/` probably does not need escaping. The idea is to escape more than what is needed, because I don't know all the special characters different shells may have. Currently, there are five different shells I need to support: `PowerShell`, `Bash`, `Dash`, `Zsh` and `CMD` (the latter does not actually have any escaping done by this plugin, because I don't know if it's possible to do it in CMD). I rather do unneeded escaping than leave out some characters that I didn't know were special characters in some case. If the escaping works the way it's meant to work, also the unnecessary escapings will disappear.

Where would the user do the decision of what characters to escape and what not? In the content they pass to the shell command, or somewhere in the settings? Somewhere else? If in the settings, then they might end up not to see all the possible values/characters they may pass to the shell command in the future. If in the content itself, then I think it's manual work that users may forgot to do, and that would also slow them down.

I'm glad to have this conversation, as I feel I can learn something new and improve this plugin. 🌞

**Reply by wealthychef1** [reply]
> > but actually it just moves the problem and creates a new special character to worry about, which is actually harder to remove because backslash is itself a special shell character.
>Sorry, I did not understand this. Can you try to explain it again?

Sure, what I mean is that if you read the man page for bash, it lists special characters for the shell.  http://mywiki.wooledge.org/BashGuide/SpecialCharacters  If you're going to escape characters, it should be those I suppose, but I don't understand what you are trying to do with this business.  


> > I think you are assuming that cleaning up all your backslashes is trivial compared to removing spaces,
> 
> Yes, I do assume that cleaning up the backslashes is trivial, as I assume your shell does it for you. But as it seems that your shell does not clean the backslashes, I'd like to know more about this situation: what is your shell and your operating system? Maybe it could help me understand why the backslashes are left uncleaned.

I think what seems to be happening is that the shell is not actually being invoked in a way that triggers interpolation.  
First, as a test, I put this on the clipboard:  `/Volumes/2_TB_SSD_external/Projects_2TB/wealthyvault/Excalidraw/Learning Excalidraw.excalidraw.md`. No backslashes there.  I see that the plugin is getting passed a string containing '\\' from the {{clipboard}} variable, even though there are no '\\' characters in the clipboard in reality.   So the problem appears to be in the definition or handling of the {{clipboard}} variable to my eyes.  ? 

![2022-05-20 14 35 12 CleanShot Obsidian](https://user-images.githubusercontent.com/4825448/169615618-0e11cdad-bf84-4e06-b570-0be1b1255617.png)

This is where the damage is done.  After that, no shell appears to be invoked, or perhaps it's being invoked in a way that tells it not to do string parsing.  
So I'm left wondering, how is bash being invoked by your code?  I don't understand yet enough of Obsidian plugin architecture to look into it... 

> 
> > but actually it just moves the problem and creates a new special character to worry about, which is actually harder to remove because backslash is itself a special shell character.
> 
> Sorry, I did not understand this. Can you try to explain it again?

I am starting to realize that since normally the shell should remove the '\' as you say, but it is not happening.  
Perhaps this example from my Terminal will be instructive:  

```bash
rcook@MacBook-Pro-2021 (wealthyvault (master)): echo "string\with backslash"
string\with backslash
rcook@MacBook-Pro-2021 (wealthyvault (master)): echo string\with backslash
stringwith backslash
```

> 
> > For most cases, I understand shell escaping and quoting just fine. As someone who has used bash and unix for years, having a program try to escape its output like this is not helpful.
> 
> Ok, good to know that you are well experienced with shell escaping 👍. Then you are well prepared to use the unescaped `!` way. I'd still like us to continue inspecting why the escaping does not work in your use case. You don't have to use the escaping if you don't want to, I'd just like to find a solution to make it work, just to learn myself and to find out if there's anything for me to improve in the plugin.
> 
> > You have to assume you know my use cases, and you can't possibly know.
> 
> Yes, the escaping that the plugin uses, does assume that the variable values are used in a certain way, i.e. not wrapped in any quotes (except in PowerShell, but that's not the case here). Apparently, there's more assumptions, that I'm yet unaware of, because you are having this problem even when you don't use any quotes. I'd like to know what it is that causes the backslashes not to disappear.

I think we are narrowing in here.  You say the {{clipboard}} is unquoted but apparently it is indeed getting quotes somehow "internally".  So `echo -n {{clipboard}}` acts the same as `echo -n "{{clipboard}}"` but it should not, agreed?   That is the real mystery, as apparently you are expecting it to act like an unquoted string, but it's getting quoted ==before== you are applying your backslash trickery.  
As a side question, why not just use double quotes?  Not that I want that either... it's just easier  

> 
> I don't have extensive experience on shell escaping myself. I wanted to come up with an escaping system that is safe to use even when the variable values contain characters the user who wrote the shell command did not expect. It's really valuable to get feedback about the escaping 🙂.
> 
> 
> Where would the user do the decision of what characters to escape and what not? In the content they pass to the shell command, or somewhere in the settings? Somewhere else? If in the settings, then they might end up not to see all the possible values/characters they may pass to the shell command in the future. If in the content itself, then I think it's manual work that users may forgot to do, and that would also slow them down.

Normally, I choose to simply escape whatever the shell is going to steal. It is always possible to craft a bad input, and the generic Shell Command plugin is in my view the wrong place to defend against this.  It should be done by each program.  In Unix, all inputs are treated as plain files... 

> 
> I'm glad to have this conversation, as I feel I can learn something new and improve this plugin. 🌞

Me too!  :-). Thanks for engaging with it.  I'm learning too.  I'm very curious why my string is getting soooo escaped.  :-)

**Reply by Taitava** [reply]
> > > but actually it just moves the problem and creates a new special character to worry about, which is actually harder to remove because backslash is itself a special shell character.
> > > Sorry, I did not understand this. Can you try to explain it again?
> 
> Sure, what I mean is that if you read the man page for bash, it lists special characters for the shell. http://mywiki.wooledge.org/BashGuide/SpecialCharacters If you're going to escape characters, it should be those I suppose, 

Can you explain more about how the escaping **moves** the problem?

I might consider narrowing the group of escaped characters, but only after I'll have more experience on shells, so not any time soon. 

> but I don't understand what you are trying to do with this business.

I'm not sure I can explain the reasoning any better than what I've tried in this discussion 🙂. I'd like to hear your proposal for a better way to pass any kind of input as literal text.

> I think what seems to be happening is that the shell is not actually being invoked in a way that triggers interpolation. First, as a test, I put this on the clipboard: `/Volumes/2_TB_SSD_external/Projects_2TB/wealthyvault/Excalidraw/Learning Excalidraw.excalidraw.md`. No backslashes there. I see that the plugin is getting passed a string containing '\' from the {{clipboard}} variable, even though there are no '\' characters in the clipboard in reality. So the problem appears to be in the definition or handling of the {{clipboard}} variable to my eyes. ?
> 
> ![2022-05-20 14 35 12 CleanShot Obsidian](https://user-images.githubusercontent.com/4825448/169615618-0e11cdad-bf84-4e06-b570-0be1b1255617.png)
> 
> This is where the damage is done. After that, no shell appears to be invoked, or perhaps it's being invoked in a way that tells it not to do string parsing.

At least the screenshot looks like the plugin is working just the way it is designed to work. The gray text is a preview of the shell command **before** execution, so there's no shell involved. Can you [set the `stdout` output channel](https://publish.obsidian.md/shellcommands/Output+handling/Output+channels#How+to+select+output+channels+for+a+shell+command) to be [Ask after execution](https://publish.obsidian.md/shellcommands/Output+handling/Output+channel+-+Ask+after+execution) and provide me a screenshot of the output after you have executed the shell command, please?


> So I'm left wondering, how is bash being invoked by your code? I don't understand yet enough of Obsidian plugin architecture to look into it...

This plugin has grown quite big by its codebase, so I'm not going to explain the whole execution pipeline, but here's the innermost part that actually executes shell commands:
https://github.com/Taitava/obsidian-shellcommands/blob/ac51071a5891cfae84eaf22f1566c1cc018c9543/src/ShellCommandExecutor.ts#L201


> I am starting to realize that since normally the shell should remove the '' as you say, but it is not happening. Perhaps this example from my Terminal will be instructive:
> 
> ```shell
> rcook@MacBook-Pro-2021 (wealthyvault (master)): echo "string\with backslash"
> string\with backslash
> rcook@MacBook-Pro-2021 (wealthyvault (master)): echo string\with backslash
> stringwith backslash
> ```

Yep, I'm aware of this. If there are quotes around the string containing a `\` backslash, the backslash won't get removed. That's why [I'm instructing not to wrap any variables in quotes when using Bash](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values#Escaping+depends+on+the+shell):
> _"Most of the time, the escape characters works quite similarly. However, on Bash and the like, you should not have escaped characters wrapped in double quotes `"` or single quotes `'`, because that will leave the escaping character `\` visible in the final result. On PowerShell, however, this does not seem to happen."_

---

> I think we are narrowing in here. You say the {{clipboard}} is unquoted but apparently it is indeed getting quotes somehow "internally". So `echo -n {{clipboard}}` acts the same as `echo -n "{{clipboard}}"` but it should not, agreed? That is the real mystery, as apparently you are expecting it to act like an unquoted string, but it's getting quoted ==before== you are applying your backslash trickery.

Are we talking about the same thing when speaking about _quoting_/_quotes_? When I speak about _quotes_, I mean double quotes `"` and/or single quotes `'`. My plugin does not append any quotes. In some cases, _quoting_ can mean the same as _escaping_, e.g. [`preg_quote()` function in PHP](https://www.php.net/manual/en/function.preg-quote.php) escapes regex patterns by adding backslashes in front of certain special characters. I mentioned the PHP function only as a terminonological example even though PHP is not used in this project.

I can agree that without quotes around the string, backslashes should **not** be visible in the output of the shell command. However, they should be visible in the preview text.

> As a side question, why not just use double quotes? Not that I want that either... it's just easier

Just wrapping variables in double/single quotes is not a safe alternative for escaping, because if a variable's value contains a double or a single quote, part of the value will jailbreak the quotes. E.g. `echo "{{!clipboard}}"` can be dangerous if `{{!clipboard}}` happens to contain e.g. `TEXT" > ImportantFile.md"`, in which case the whole shell command would become `echo "TEXT" > ImportantFile.md""`. This would make the command to write text into a file, possibly overwriting an important file, which was not the original purpose of the command.



> > I don't have extensive experience on shell escaping myself. I wanted to come up with an escaping system that is safe to use even when the variable values contain characters the user who wrote the shell command did not expect. It's really valuable to get feedback about the escaping 🙂.
> > Where would the user do the decision of what characters to escape and what not? In the content they pass to the shell command, or somewhere in the settings? Somewhere else? If in the settings, then they might end up not to see all the possible values/characters they may pass to the shell command in the future. If in the content itself, then I think it's manual work that users may forgot to do, and that would also slow them down.
> 
> Normally, I choose to simply escape whatever the shell is going to steal. It is always possible to craft a bad input, and the generic Shell Command plugin is in my view the wrong place to defend against this. It should be done by each program.

Even if _it is always possible to craft a bad input_, I'm trying to learn more and find out if there are solutions for bad input cases I haven't been able to deal with in the plugin previously. I don't agree that the fact that there may be unnoticed security holes would mean that the plugin should not defend against **known** security holes. If it would not be done by the plugin, where would it be done then? 🙂 Can you provide a concrete example, please? 🙂 

> In Unix, all inputs are treated as plain files...

What do you mean by this? 🙂

**Reply by wealthychef1** [reply]
I think it will help to ask you directly something that I'm trying to escape where your system is giving me trouble, or where I don't understand how to use it correctly.  
### Example
How do I do this using your quoting system, i.e. without a `!` symbol?  

```
 bash echo -n [my label](<file://{{clipboard}}>)
```

I want all those ugly special characters in there, and I don't want any backslashes added.  
So if the clipboard is 
`/Planning and Organization/2014-10-31 The great stories of my life, and those to come.md` then I want the output on the page to be literally 
```
[my label](<file:///Planning and Organization/2014-10-31 The great stories of my life, and those to come.md>)
```

But what comes out is
```
/bin/bash: -c: line 0: syntax error near unexpected token `('
/bin/bash: -c: line 0: `bash echo -n [clipboard contents](<\.\/Planning\ and\ Organization\/2014\-10\-31\ The\ great\ stories\ of\ my\ life\,\ and\ those\ to\ come\.md\\n>)'
```

So I need to escape the special characters in my string.  
One approach is 
`bash echo -n '[my label](<'{{clipboard}}'>)'`, selectively unquoting the macro so you can escape it for me.  
But it is much easier now to just do this: 
`bash echo -n '[my label](<file://{{!clipboard}}>)'`

This way, I know the single quotes protect the entire clipboard from the shell's interpolation.  If I leave it outside the quotes, I am exposed to the problem of having to trust your quoting rules implicitly.  If I don't, I understand what's up.  

I think this use case walks you through why I prefer you do not quote any characters for me and just let me escape what needs to be escaped as a knowledgable bash user.  Here I just use single quotes.  Even if you perfectly imitate the bash quoting rules, inserting a forest of backslashes is not a great approach in my view!  

I now feel I have a good set of solutions and understand how to properly quote things with Shell Commands, so goals accomplished here for me!  Thanks!

**Reply by wealthychef1** [reply]
> > In Unix, all inputs are treated as plain files...

> What do you mean by this? 

https://en.wikipedia.org/wiki/Everything_is_a_file

**Reply by Taitava** [reply]
> One approach is
> `bash echo -n '[my label](<'{{clipboard}}'>)'`, selectively unquoting the macro so you can escape it for me.

Yes, this is the way I personally use.

> But it is much easier now to just do this:
> `bash echo -n '[my label](<file://{{!clipboard}}>)'`
> 
> This way, I know the single quotes protect the entire clipboard from the shell's interpolation.

Here the protection given by single quotes does not cover everything. If `{{!clipboard}}` happens to contain a single quote `'` character, it will close your single quotes and the rest of the value of `{{!clipboard}}` can be dangerously interpolated by the shell.

> If I leave it outside the quotes, I am exposed to the problem of having to trust your quoting rules implicitly. If I don't, I understand what's up.

I can very well understand this point. My escaping mechanism cannot be quaranteed to work 100% correct in every situation. I just cannot claim to have all the needed expertise.

I guess this is a choice every user must make themselves: either trust the self-made escaping system, or trust the input values do not contain single quotes.

This is a really valuable discussion. 🙂👍

**Reply by wealthychef1** [reply]
> > One approach is
> > `bash echo -n '[my label](<'{{clipboard}}'>)'`, selectively unquoting the macro so you can escape it for me.
> 
> Yes, this is the way I personally use.
> 
> > But it is much easier now to just do this:
> > `bash echo -n '[my label](<file://{{!clipboard}}>)'`
> > This way, I know the single quotes protect the entire clipboard from the shell's interpolation.
> 
> Here the protection given by single quotes does not cover everything. If `{{!clipboard}}` happens to contain a single quote `'` character, it will close your single quotes and the rest of the value of `{{!clipboard}}` can be dangerously interpolated by the shell.

That is a good point, but maybe I WANT the shell to help me, like to expand a tilde or do other tricky things.  The good news is that with the ! escape, users can make the choice for themselves.  
Escaping stuff in the shell is a lifelong adventure it seems.  :-) 

> 
> > If I leave it outside the quotes, I am exposed to the problem of having to trust your quoting rules implicitly. If I don't, I understand what's up.
> 
> I can very well understand this point. My escaping mechanism cannot be quaranteed to work 100% correct in every situation. I just cannot claim to have all the needed expertise.
> 
> I guess this is a choice every user must make themselves: either trust the self-made escaping system, or trust the input values do not contain single quotes.

Yes and the more clearly your rules for what gets escaped are spelled out in the docs, the more easily users can make informed choices!  

> 
> This is a really valuable discussion. 🙂👍

Totally agree!  Now this is out there for others to see as well.  I feel well-armed to use your tool better.

**Reply by Taitava** [reply]
> That is a good point, but maybe I WANT the shell to help me, like to expand a tilde or do other tricky things. The good news is that with the ! escape, users can make the choice for themselves.

True, there might be cases when the special characters are wanted to be used without escaping 👍.

> Escaping stuff in the shell is a lifelong adventure it seems. :-)

It certainly is! 😎 

> > > If I leave it outside the quotes, I am exposed to the problem of having to trust your quoting rules implicitly. If I don't, I understand what's up.
> > 
> > 
> > I can very well understand this point. My escaping mechanism cannot be quaranteed to work 100% correct in every situation. I just cannot claim to have all the needed expertise.
> > I guess this is a choice every user must make themselves: either trust the self-made escaping system, or trust the input values do not contain single quotes.
> 
> Yes and the more clearly your rules for what gets escaped are spelled out in the docs, the more easily users can make informed choices!

I'd like to receive feedback about this documentation page regarding escaping: https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values

A quote from the documentation page that contains the core definition of characters that SC escapes in variable values:
> _"A character is considered special by SC if it's **not** an alphabet in ASCII character set `a`-`z`/`A`-`Z`, not a number `0`-`9`, and not an underscore `_`."_

I would be glad to hear about any improvements you (or others) might have in your mind regarding the page! 🙂 


> > This is a really valuable discussion. 🙂👍
> 
> Totally agree! Now this is out there for others to see as well. I feel well-armed to use your tool better.

Nice to hear! 🌞 Yea, this is valuable for others, too.

**Reply by wealthychef1** [reply]
> I'd like to receive feedback about this documentation page regarding escaping: https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values

I would just say that you should make it more obvious how to use the ! method to escape your escapes, instead of trying to persuade users that it's the wrong choice for them.  It's almost like you are trying to hide it.  Don't be afraid users will disagree with your opinion, let them figure out which they prefer.  If you're lucky, EVERYONE will use ! and you won't have to worry about escapes... 

In my view for my case as an experienced unix user, you should recommend the user of ! because it allows a knowledgable user to do things they way they normally do things.  

Cheers and thanks for all the fish!

**Reply by Taitava** [reply]
> I would just say that you should make it more obvious how to use the ! method to escape your escapes

Good feedback! 👍 I added a tip about disabling escaping [to the top of the page](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values#Escaping+special+characters+in+variable+values). Any comments? 🙂 

> Don't be afraid users will disagree with your opinion, let them figure out which they prefer.

I'm not afraid that users might disagree, I just want to guarantee that users know what they are doing before disabling the escaping 🙂. You certainly seem to know what you are doing and what you want.

I prefer the following: Users might notice excess `\` or `` ` `` escaping characters and then they go read the documentation or come here to find out how to prevent it. Sure, it's some work for the user and a bit annoying perhaps, but I like it more than seeing users reporting how their variable values accidentally overwrote important files or executed wrong shell commands.

> If you're lucky, EVERYONE will use ! and you won't have to worry about escapes...

Then I would worry about everyone disabling the escaping 🙂. But if they do, then they do - at least they are warned. There are valid use cases for unescaped values, I'm not denying that, all I want is users to know when they are safe to use and when not.

> Cheers and thanks for all the fish!

Thank you! 🌞

**Reply by wealthychef1** [reply]
Thanks again.  Your addition to the documentation is perfect for my mind.  To be honest, it's just hard to use Obsidian still, as powerful as it is, most users would not spend the time I have trying to understand it.  Shell Commands is a huge win for someone like me obviously.  You have done a great job in providing this plugin.  I can't stress how useful I find it.  Thank you!

**Reply by Taitava** [reply]
Thank you a lot for your kind feedback! 😃

**Comment by Taitava** [comment]
## I moved this conversation from _Ideas_ to _Q & A_.
